packages/core/src/sync-store/index.ts

```diff
@@ -1089,13 +1090,102 @@ export const createSyncStore = ({
     if (requests.length === 0) return [];
 
+    // Optimized fast path for high number of `requests` using a range of block numbers
+    // rather than querying each request individually.
+
+    const blockNumberRequests: (EIP1193Parameters & { blockNumber: number })[] =
+      [];
+    const nonBlockNumberRequests: EIP1193Parameters[] = [];
+    for (const request of requests) {
+      const blockNumber = extractBlockNumberParam(request);
+
+      // Note: "latest" is sorted into `nonBlockNumberRequests`
+      if (isHex(blockNumber)) {
+        blockNumberRequests.push({
+          ...request,
+          blockNumber: hexToNumber(blockNumber),
+        });
+      } else {
+        nonBlockNumberRequests.push(request);
+      }
+    }
+
     const requestHashes = requests.map((request) =>
       crypto
         .createHash("md5")
         .update(toLowerCase(JSON.stringify(orderObject(request))))
         .digest("hex"),
     );
 
+    if (blockNumberRequests.length > 100) {
+      const minBlockNumber = Math.min(
+        ...blockNumberRequests.map((request) => request.blockNumber),
+      );
+      const maxBlockNumber = Math.max(
+        ...blockNumberRequests.map((request) => request.blockNumber),
+      );
+
+      const result = await Promise.all([
+        database.syncQB.wrap({ label: "select_rpc_requests" }, (db) =>
+          db
+            .select({
+              request_hash: PONDER_SYNC.rpcRequestResults.requestHash,
+              result: PONDER_SYNC.rpcRequestResults.result,
+            })
+            .from(PONDER_SYNC.rpcRequestResults)
+            .where(
+              and(
+                eq(PONDER_SYNC.rpcRequestResults.chainId, BigInt(chainId)),
+                gte(
+                  PONDER_SYNC.rpcRequestResults.blockNumber,
+                  BigInt(minBlockNumber),
+                ),
+                lte(
+                  PONDER_SYNC.rpcRequestResults.blockNumber,
+                  BigInt(maxBlockNumber),
+                ),
+              ),
+            ),
+        ),
+        nonBlockNumberRequests.length === 0
+          ? []
+          : database.syncQB.wrap({ label: "select_rpc_requests" }, (db) =>
+              db
+                .select({
+                  request_hash: PONDER_SYNC.rpcRequestResults.requestHash,
+                  result: PONDER_SYNC.rpcRequestResults.result,
+                })
+                .from(PONDER_SYNC.rpcRequestResults)
+                .where(
+                  and(
+                    eq(PONDER_SYNC.rpcRequestResults.chainId, BigInt(chainId)),
+                    inArray(
+                      PONDER_SYNC.rpcRequestResults.requestHash,
+                      nonBlockNumberRequests.map((request) =>
+                        crypto
+                          .createHash("md5")
+                          .update(
+                            toLowerCase(JSON.stringify(orderObject(request))),
+                          )
+                          .digest("hex"),
```

Comment:
I think we could avoid computing again the requests hashes for those request as it was computed just before already for all requests.
Would save some CPU.